Design: Nano Contract integration within Wallet Mobile

alexruzenhack opened the pull request:
### Acceptance Criteria

- Show registered Nano Contracts
- Show Nano Contract's transactions
- Show Nano Contract's state
- Show Nano Contract transaction item details
- Sign Nano Contract transaction by Wallet Connect

See [design rendered](https://github.com/HathorNetwork/rfcs/blob/964abc6a48b3808ce30fc19a64feec157f506d67/projects/hathor-wallet-mobile/0002-nano-contract-integration.md).

andreabadesso commented on projects/hathor-wallet-mobile/0002-nano-contract-integration.md:
```diff
@@ -0,0 +1,476 @@
+Upon activation, the wallet initiates a restart process, during which Nano Contract transactions for the current wallet are loaded. Users are visually informed of this process through a progress indicator and are returned to the *Nano Contract Settings* screen upon completion.
+
+>[!NOTE]
+In addition to enabling the *Nano Contract* feature, basic usage data such as feature activation frequency and duration of use may be collected for monitoring purposes. For more details on data collection and push notification support, see [Basic Data Gathering and Push Notification Support](#basic-data-gathering-push-notification-support).
```
I don't think we should intentionally collect any kind of data for monitoring purposes

alexruzenhack replied:
I think there is no problem to get anonymized data with user consent. We may want to know the feature adoption statistics, specially to prioritize our efforts. Unleash is not reliable to get adoption statistics.

alexruzenhack replied:
I understand your concerns, they are super valid. This ins't an easy topic. I believe it would be beneficial to provoke this discussion with the team and possibly involve the community in the process. We will spend some effort in the near feature in UI and UX improvements and do it without data will really throw us in the dark. By discussing this topic we will at least discover and map our boundaries.

alexruzenhack replied:
I've dicussed this topic with @pedroferreira1 and we come to a conclusion that it is the realm of product development and @trondbjoroy should lead this.

andreabadesso commented on projects/hathor-wallet-mobile/0002-nano-contract-integration.md:
```diff
@@ -0,0 +1,476 @@
+
+This screen is dedicated to displaying a list of *Nano Contract Transaction* items associated with a selected *Nano Contract*. After a user selects a *Nano Contract* from the *Nano Contract List* component, they are directed to this screen, where all transactions involving the chosen contract are listed.
+
+The header section of this screen provides essential state information about the selected *Nano Contract*, including the full *Nano Contract ID*, *Blueprint Name*, and a breakdown of each *Token Balance*. This header is designed to be both expandable and collapsible: in its expanded form, it reveals all state details, while in its collapsed form, it displays only the *Nano Contract ID* for a more ...
```
Is the blueprint name unique in the network?

Should we also display an URL to the explorer?

alexruzenhack replied:
> Is the blueprint name unique in the network?

No, the name isn't unique.

> Should we also display an URL to the explorer?

We can have this functionality at this screen, sure.

andreabadesso commented on projects/hathor-wallet-mobile/0002-nano-contract-integration.md:
```diff
@@ -0,0 +1,476 @@
+
+For *sent* and *received* transaction items, the header displays *Token Amount* and *Token Symbol*, with an optional label *Amount* underneath them.
+
+The body section includes:
```
Should we also include information about the method call? Like which parameters were passed and the method call result?

alexruzenhack replied:
> Should we also include information about the method call?

Yes, we can include the arguments called.

> Like which parameters were passed and the method call result?

Public methods are always void.

alexruzenhack replied:
I've discussed this with Pedro and we came to the conclusion to redirect users to explorer website on the trasaction page, where a user will be able to see all the details.

andreabadesso commented on projects/hathor-wallet-mobile/0002-nano-contract-integration.md:
````diff
@@ -0,0 +1,476 @@
+    A -->|This Flow| C[Nano Contract Registration]
+```
+
+### Screen: Nano Contract Registration
````
I missed some context on what is a nano contract registration, can you add a short description on what registering a nano contract means?

alexruzenhack replied:
Sure! I'll add it.

By registering a Nano Contract the wallet can load its transactions.

andreabadesso commented on projects/hathor-wallet-mobile/0002-nano-contract-integration.md:
```diff
@@ -0,0 +1,476 @@
+	- Include detailed transaction information for user review:
+		- *Blueprint ID*, *Blueprint Method*, *Caller*, *Nano Contract ID* (if applicable), *Action List*, and *Argument List*.
+	- Include detailed information for each *Action List* item:
+		- *Type*, *Token* and *Data*.
```
What is data? Is it different for different nano contracts?

alexruzenhack replied:
Data encompass "address" and "amount" for an action type "withdrawal". If it is a "deposit" type, it will have only the "amount". In the end it will be translated to UTXOs.

andreabadesso replied:
Are there other action types? Can you add more context in the design?

alexruzenhack replied:
No, there aren't more types than that. Sure, I will add more context in the design.

andreabadesso commented on projects/hathor-wallet-mobile/0002-nano-contract-integration.md:
```diff
@@ -0,0 +1,476 @@
+	- Include detailed information for each *Argument List* item:
+		- *Type* and *Value*
+	- Provide clear options for user approval or rejection of the transaction.
+	- Upon tapping *Approve*, initiate the processing with appropriate loading feedback.
```
Since this might appear on any screen at any time, should we add a secondary confirmation modal to make sure the user intentionally clicked approve? I feel that he might misclick the button during normal usage of the app

alexruzenhack replied:
I don't like second modals, but we can use another strategy like tap in in a text "Take action on this request" and after it the buttons shows up. Change the action order. Set the primary button to "Reject" or "Dismiss".

anafigueired replied:
Other valid option is an Error Recovery strategy, such as undoing actions or restoring previous states.
This kind of strategy provides users with a clear and easy way to recover from errors, increasing the app safety.
If its something possible, we should consider as well: having only the reject or dismiss buttons, but an "undo previous action" message after.

andreabadesso replied:
An undo button wouldn't work here because after the signed data is sent to the dApp, there is no going back

I like @alexruzenhack's suggestion tho

andreabadesso commented on projects/hathor-wallet-mobile/0002-nano-contract-integration.md:
```diff
@@ -0,0 +1,476 @@
+	- The modal must render centrally on top of the current screen, ensuring clear visibility and focus.
+	
+- **Functionality**:
+	- Include detailed transaction information for user review:
```
Maybe we should warn users that they should reject the modal if they don't fully understand the parameters

alexruzenhack replied:
Good ideia!

anafigueired replied:
We could also have a "read more" link after the messages, if we have some documentation about that, so if the user is still confused, they could seek for mor information before making a decision.

alexruzenhack replied:
This is also a good ideia. I can mention it on the design. However, it is good to stress that this design is an experiment and I wrote it without tight design specification because I want to create room for your work. That said, I will add it as optional, and you decide in the end. What do you think?

andreabadesso commented on projects/hathor-wallet-mobile/0002-nano-contract-integration.md:
```diff
@@ -0,0 +1,476 @@
+## Load Nano Contract History
+
+##### When execute
+- It should happen everytime a new *Nano Contract Transaction* containing a caller from user's wallet is processed on `sagas/wallet.js@handleTx`.
```
Should we also reload the history when a nano contract which the user has registered state mutated?

alexruzenhack replied:
Well, the only way to mutate nano contract's state is by a transaction. Therefore, this is already capturing the state mutation.

But when you say "reload the history" are you referring to a re-render of the screen?

andreabadesso replied:
> Well, the only way to mutate nano contract's state is by a transaction. Therefore, this is already capturing the state mutation.

Oh, ok!

andreabadesso commented on projects/hathor-wallet-mobile/0002-nano-contract-integration.md:
```diff
@@ -0,0 +1,476 @@
+
+- `blueprintId`: unique identification of the *Blueprint*
+- `method`: *Blueprint*'s method name
+- `caller`: wallet address representing method's *Caller*
```
Can you give more context here on the nano contracts specific parameters or point to the documentation?

alexruzenhack replied:
See the following [signature](https://github.com/HathorNetwork/hathor-wallet-lib/blob/50dfe66e720b481619e77558ddfc83e7096bae27/src/new/wallet.js#L2619):
```ts
async createAndSendNanoContractTransaction(blueprintId, method, address, data, options = {})
```

andreabadesso commented on projects/hathor-wallet-mobile/0002-nano-contract-integration.md:
```diff
@@ -0,0 +1,476 @@
+3. If rejected, return to dapp an error message
+4. If approved, ask user's Pin Code
+5. Create and send transaction
+6. Return to dapp a success message 
```
In this specific example, shouldn't the user approve the use of his utxos?

alexruzenhack replied:
Oh! Yes! I didn't stressed this part because after the request be approved almost immediately the PIN will be required, but I will make it clear.

alexruzenhack replied:
Humm... I see this operation like the send transaction. The use didn't see the UTXOs, the wallet chose the UTXOs to be used. I mimic this behavior there:

```
4. If approved, ask user's Pin Code
```

alexruzenhack replied:
Do you believe this is necessary? For this we would need a method on the wallet lib side just to mount the transaction and another just to sign and publish, other than we would need more screens in this project to carry on this behavior.

andreabadesso replied:
No, you're right, we don't ask for confirmation on the UTXOs the lib selected in the send transaction method, so we shouldn't here either

anafigueired commented:
**General UX comments:**

**- On Nano Contracts Settings Flow:**
1. Post loading error: Map the erros to create an error message with action so the user doesn't get frustrated. E.g.: Try again | Go back (Double check with UX the messages in order to start implementing standard error messages during the app flow).

2. Success message: Offer an extra action for the user to see the list of Nano Contracts (this follow an easy way to navigate with autonomy strategy).

3. Suggestion: In the toggle interface, include an informative component, such as an information box, to provide a brief explanation about the functionality and a link to more information outside the app. This not only fills the screen but also offers users a contextual view (enhancing safety, information and autonomy for the user).

4. Suggestion for (maybe) future: Is it worth onboarding the functionality? When the user activates, a modal will appear with 3 or 4 screen scrolls showing how simple it is to use, so they know what to expect.


**- On Main Screen Flow:**

1. Regarding the way of writing, the title "Nano Contract List" can really be optimised. The most important word, which communicates functionality, is "Contract" and not "Nanos". I suggest testing "Nano Contracts" and evaluating the readability on the screen.

2. Always use the same term on all screens to ensure standardisation (e.g. not having Nano on one, and Contracts on another).

3. It is worth consulting marketing regarding the naming, because it also influences branding and our position in the market (I am gonna have a meeting with mkt about that, so we can talk later about this subject). 


**- On Modal Flow:**

1. I have commented below the reviewer's comments on specific part about "Error Recovery strategy".

alexruzenhack commented:
> * Post loading error: Map the erros to create an error message with action so the user doesn't get frustrated. E.g.: Try again | Go back (Double check with UX the messages in order to start implementing standard error messages during the app flow).
> *

I believe we don't need to specify all possible errors, but specify error categories. In this respect we have only 3 categories:
- **Validation Error**: it happens when a user inputs an invalid data in a form. The error must appear in the input context.
- **Feedback Error**: it happens when some error happens during processing but don't impact wallet's state. The error must appear as a feedback modal with an error message, and the user can try the action again or dismiss.
- **Crash Error**: it happens when some error happens during processing but impacts wallet's state or its operation. The error must appear as a feedback modal with an error message and a message informing the user that the wallet need to be restarted. In addition, we should give an option for the user to send the error to our team for further analysis.

> 2\. Success message: Offer an extra action for the user to see the list of Nano Contracts (this follow an easy way to navigate with autonomy strategy).

I believe it is a nice addition, indeed. Let's it open for your discretion when you will be designing the screens and flows.

> 3\. Suggestion: In the toggle interface, include an informative component, such as an information box, to provide a brief explanation about the functionality and a link to more information outside the app. This not only fills the screen but also offers users a contextual view (enhancing safety, information and autonomy for the user).

I've added this suggestion.

> 4\. Suggestion for (maybe) future: Is it worth onboarding the functionality? When the user activates, a modal will appear with 3 or 4 screen scrolls showing how simple it is to use, so they know what to expect.

I'm mentioning this suggestion in the [Feature onboarding](https://github.com/HathorNetwork/rfcs/blob/7ada93db3f9287873e85e278f0aa773c6b462200/projects/hathor-wallet-mobile/0002-nano-contract-integration.md#feature-onboarding) topic.

> 3\. It is worth consulting marketing regarding the naming, because it also influences branding and our position in the market (I am gonna have a meeting with mkt about that, so we can talk later about this subject).

Sure! I believe this is a great contribution you can bring to the table in your design outcome.

> 1. I have commented below the reviewer's comments on specific part about "Error Recovery strategy".

@andreabadesso has [replied here](https://github.com/HathorNetwork/rfcs/pull/72#discussion_r1432704713) about it. Just stressing what he have said, there isn't recovery for transaction. We can't rollback transactions to restore previous wallet balance state. All the transactions are immutable and unidirectional. That's why the user MUST be certain what it is doing. On the design side we should help the user applying the ["Clear Signing" principal](https://www.ledger.com/blog/clear-sign-your-worries-away) and other strategies to make mistakes difficult.

hBulldozer commented:
**Suggestion for Nano Contract Display Design**

Hi @alexruzenhack, I am Bulldozer a developer testing Nano Contracts and building a DeFi Protocol dozer.finance

I'm following up on the wallet connect design discussion I had with @andreabadesso, where we brainstormed some ideas for your nano contract display design. He suggested I share them here too, so here goes!

**Simplifying Contract Call Display:**

I believe displaying all contract calls within the token transaction history would be more user-friendly. This would consolidate a user's spending and earnings per token, regardless of transaction type. For example, with 1000 HTR and 200 DZR, and two transfers, one swap, and one staking transaction, the history might look like:

```
Token List
810 HTR 
100 DZR 

Click for details(HTR): 
*...
* Transaction 1: +10 HTR
* Transaction 2: -100 HTR
* Contract call: -100 HTR, +50 DZR

Click for details(DZR): 
*...
* Contract call: +50 DZR, -100 HTR
* Contract call: -150 DZR
```

This concise format provides a clear overview of all transactions, including contract calls.

**Nano Tab for Detailed History:**

The "Nanos" tab could then offer a more detailed breakdown of contract call history, including contract address, blueprint name, and transaction details. For example:

```
Contract Calls (History)

* Contract call: -100 HTR, +50 DZR
* Contract call: -150 DZR

Click for details:
* Contract address: 0x1234567890abcdef
* Blueprint name: Swap
* Transaction details:
    * Token 1: HTR
    * Token 2: DZR
    * Amount 1: -100
    * Amount 2: +50
```

This deeper level of information can be helpful for troubleshooting and debugging.

**User Interaction with Contracts:**

I understand that this approach wouldn't allow direct interaction with new contracts from the wallet. However, considering the complexity of calling contracts without code knowledge, perhaps building simple interfaces for specific, user-friendly blueprints could be explored. And for the more complex one the dApps instantiating these contracts will build an interface to help users to call the methods with correct values and the user just sign with wallet connect.

**Variable Pattern for Locked Funds:**

Finally, if blueprints follow a consistent data structure for user funds, displaying locked funds within interacted contracts directly in the wallet might be possible. For example, using a pattern like `state[]=user.deposits.a'{{Address}}'` in the contract the user has in his history.

I hope these suggestions are helpful. I'm eager to hear your thoughts and discuss them further!

Thanks,

alexruzenhack commented:
@hBulldozer Thank you for your collaboration. I incorporated some of your suggestions in the design, see bellow:

> **Simplifying Contract Call Display:**

Today we already show Nano Contract transactions and the token balance in the transaction. I have specified in the design a suggestion for the Designer to mark Nano Contract transactions, and I have included a specification of Nano Contract information on those transactions.

See: [Component Token List](https://github.com/HathorNetwork/rfcs/blob/015f77558898c1f3e540faa5d14bfadea13d5008/projects/hathor-wallet-mobile/0002-nano-contract-integration.md#component-token-list), Component Token Transactions and Component Token Transaction Details.

> **Nano Tab for Detailed History:**

I see value in this suggestion, therefore I added it as an [alternative](https://github.com/HathorNetwork/rfcs/blob/015f77558898c1f3e540faa5d14bfadea13d5008/projects/hathor-wallet-mobile/0002-nano-contract-integration.md#alternative-component-nano-contract-transaction-list). @pedroferreira1 has to give its review yet, let's wait his contribution as well.

> **User Interaction with Contracts:**

Nice that you have mentioned it. I've discussed this possibility with @pedroferreira1 in previous talks. I'm registering it in the alternatives because we plan to work on it in the future, see [Specialized blueprint interface](https://github.com/HathorNetwork/rfcs/blob/0d8e6e97072ce486f26b1675338567ed36347821/projects/hathor-wallet-mobile/0002-nano-contract-integration.md#specialized-blueprint-interface).

> **Variable Pattern for Locked Funds:**

I didn't understood this one. Can you provide me more context with examples?

anafigueired reviewed (commented):
The new suggestions seems fine. 
Only one comment about Specialized blueprint interface and others complex user interactions that may appear: I believe they are important but should keep for the future, so we can see how deep is going to be the level of interaction from the firsts users to the feature before adding more complexibility.

alexruzenhack commented on projects/hathor-wallet-mobile/0002-nano-contract-integration.md:
```diff
@@ -0,0 +1,591 @@
+[unleash]: #unleash
+
+First, we should enable the user's device in the `nano-contract.rollout` feature flag on Unleash. For this feature flag we use the default variant with 3 strategies:
+1. UserIds: this first strategy is dedicated for internal testers
+2. UserIds: this second strategy is dedicated for external testers
```
In the code, nothing changes. They are logical groups. I'm doing this to increase our control and convenience. Sometimes we want to turn-off the feature for everybody except for ourselves and by having these groups segregated we can achieve this with more precision and less work, it would be a no brainer.

alexruzenhack commented on projects/hathor-wallet-mobile/0002-nano-contract-integration.md:
```diff
@@ -0,0 +1,591 @@
+### Screen: Nano Contract Settings
+[screen-nano-contract-settings]: #screen-nano-contract-settings
+
+The *Nano Contract Settings* screen features a "Nano Contract" toggle button. Users can tap this button to activate the nano-contract feature. The toggle button should have a context description for what it actually does "Enables the view of a Nano Contract list, a list of transactions per Nano Contract and the possibility to register and unregister a Nano Contract".
```
> In that case, the better UX for me would be to ask the user to configure his wallet before loading anything. With the proposed solution, we require the user to load his wallet fully, notice that nano transactions are not loaded, find the option in the settings, enable it, and then load everything again. This might lead to lot of questions and complaints that the nano transactions are not showing.

Let me break this down.

>In that case, the better UX for me would be to ask the user to configure his wallet before loading anything.

You mean during the wallet setup? After select import or create a wallet?

>[...] notice that nano transactions are not loaded

No, the nano contract transactions are being loaded all the time. I have tested it. What is not happening is the balance calculation per Nano Contract ID and other index activity that may need to happen related to Nano Contract.

>[...] find the option in the settings, enable it, and then load everything again.

Yes, this would be necessary. However, we can use the Push Notification strategy to draw user's attention by using a "pop-up" modal in the first initialization of the wallet after it be allowed to access the feature. We can even improve this process by driven the user to an onboarding experience.

>[...] This might lead to lot of questions and complaints that the nano transactions are not showing.

Nano Contract transactions are already being shown in the wallet _as is_. By enabling the feature the user would only receive a better visualization, a more contextualized data and the option to engage with Nano Contract by using the Wallet Connect. This last topic is more problematic, indeed. However, I believe in the opposite effect, if we rollout the feature without user participation, it will expose the user to new capabilities as an unexpected move, and the consequence would be the doubt and complaints. On the other hand, if we give users the power to decide, we also gain a change to guide the process using an onboarding experience, and this can greatly remove noise.

>I prefer that we have the feature toggle in the unleash only and discuss this new feature of having a full configuration of the wallet in the future.

Give me your vision about the previous reply first. But sure, we can turn this feature toggle a "Future Possibility".

alexruzenhack commented on projects/hathor-wallet-mobile/0002-nano-contract-integration.md:
```diff
@@ -0,0 +1,591 @@
+## Feature onboarding
+[feature-onboarding]: #feature-onboarding
+
+Presents an onboarding with 3 or 4 views explaining the Nano Contract feature, its functionalities and how to use. The onboard can appears after the toggle button activation and in the end ask the user to proceed with wallet reloading.
```
Well, I don't see any problem to have both, but the nature of either solutions are very different. Thinking only in the user's experience during the wallet usage, I prefer to keep continuity of the experience, therefore the onboarding better suits this principal. I believe it would be more fun and less expensive in terms of attention to users.

alexruzenhack replied:
@anafigueired Can you give your view on this matter? You were the one that proposed it.

alexruzenhack commented on projects/hathor-wallet-mobile/0002-nano-contract-integration.md:
```diff
@@ -0,0 +1,591 @@
+This component displays a list of registered *Tokens* on the *Dashboard* screen, allowing users to select and interact with each token. By clicking on a token item the user is navigated to the *Token Transactions* component.
+
+>[!NOTE]
+>The actual implementation name for the "Token List" component is `TokenSelect`, which is found in the *Dashboard Screen*. It is more convenient to call "Token List" in this design because the component presents a list of tokens, which can be selected, but the selection is just a possible action, and doesn't convey the shape of component when rendered.
```
I see. Thanks for the context! I believe we can improve the semantics of components and do some refactoring. Do you think this worths a project, or it can be done as a KTLO?

alexruzenhack commented on projects/hathor-wallet-mobile/0002-nano-contract-integration.md:
```diff
@@ -0,0 +1,591 @@
+- **Functionality**:
+	- Implement an infinite scrolling view to display a potentially large number of nano contracts.
+	- Make each list item actionable, leading to the **Nano Contract Transactions** screen upon selection.
+	- Each item list item must contain *Nano Contract ID* and *Blueprint Name* information.
```
I didn't get it. Why show the caller address here? My vision for this item is to function as an aggregator and it includes aggregate the caller addresses. What am I missing?

alexruzenhack commented on projects/hathor-wallet-mobile/0002-nano-contract-integration.md:
```diff
@@ -0,0 +1,591 @@
+### Screen: Nano Contract Transactions
+[screen-nano-contract-transactions]: #screen-nano-contract-transactions
+
+This screen is dedicated to displaying a list of *Nano Contract Transaction* items associated with a selected *Nano Contract*. After a user selects a *Nano Contract* from the *Nano Contract List* component, they are directed to this screen, where all transactions involving the chosen contract are listed.
```
Sure you right. I will add this point.

alexruzenhack replied:
Another issue. For this view I would like to extend this argument or yours:
https://github.com/HathorNetwork/rfcs/pull/72#discussion_r1453650797

>The user is able to register the same nano contract ID with different addresses

Why should we couple Contract ID with Address in the registration process? This design only takes in consideration the Contract ID. What am I missing?

alexruzenhack replied:
Understood. Design updated.

alexruzenhack commented on projects/hathor-wallet-mobile/0002-nano-contract-integration.md:
```diff
@@ -0,0 +1,591 @@
+
+This component provides detailed information for each transaction item type: *sent*, *received*, and *called*.
+
+For *sent* and *received* transaction items, the header displays *Token Amount* and *Token Symbol*, with an optional label *Amount* underneath them.
```
Oh! You are reading this with the same lens as Bulldozer, this is interesting. We can discuss it together. But let me present my view again.

In the previous section "Component: Nano Contract Transaction List", I say:
>Each item [of the list] represents a token balance

In this section a say that there are 3 types of items: "sent", "receive" and "called". But these types are applied not per transaction, but per token balance. I mean, for each transaction we can have many items, one per token balance representing the types.

I present the following case:

>[!TIP]
> Screen: Main
> Component Active: Nano Contract List
> View:
> - 2c68056 (select)
> - f41de38
> - aaa0f8a
> - efdadf4
> 
> [User selects the Nano Contract 2c68056]

>[!TIP]
> Screen: Main
> Component Active: Nano Contract Transactions
> State: Nano Contract ID = 2c68056
> View:
> - Header: _metadata information_
> - Body: Nano Contract Transaction List
> 
> [On Body the wallet renders the following component]

>[!TIP]
> Component: Nano Contract Transaction List
> State: Nano Contract ID = 2c68056
> View:
> - You *received* `20 BBB` from Swap [metadata: transaction = cdcb0ac]
> - You *sent* `10 AAA` to Swap [metadata: transaction = cdcb0ac]
> - You *called* `Swap` on Swap [metadata: transaction = cdcb0ac]
> - You *sent* `100 BBB` to Swap [metadata: transaction = a3ba49b]
> - You *sent* `100 AAA` to Swap [metadata: transaction = a3ba49b]
> - You *called* `Initialize` on Swap [metadata: transaction = a3ba49b]

As you can see, on "Nano Contract Transaction List" there are many Nano Contract transactions, however I'm presenting all the token balances plus calls, the items, in one shot. The items are sorted by their respective transactions, as you can see in the metadata annotation.

I did it for one reason: a nano contract transaction can have many tokens, but also can have many arguments for the call. Arrange everything in one view like in the details can lead to poor UI/UX. On the other hand, using this method, I can show in the details piece by piece, or item by item of a Nano Contract transaction.

If you find this bad, an argue we need an actual transaction list for users to select, we can provide one e use this component to show the transaction items after user selects a transaction, let's say "cdcb0ac", and this component would show filtered items for this transaction selected.

alexruzenhack replied:
I updated the design. See the section "Screen: Nano Contract Transaction".

alexruzenhack commented on projects/hathor-wallet-mobile/0002-nano-contract-integration.md:
```diff
@@ -0,0 +1,591 @@
+
+Taping on *Public Explorer* navigates the user to the transaction page in the explorer website.
+
+###### Blueprint Method Arguments
```
>I didn't understand this screen. How is it shown? The previous one is a modal over the screen, this one would be a modal over a modal?

Oh, no. This one is a section in the previous modal.

alexruzenhack replied:
I beleive to have simplified in the new review version.

alexruzenhack commented on projects/hathor-wallet-mobile/0002-nano-contract-integration.md:
```diff
@@ -0,0 +1,591 @@
+
+### Screen: Nano Contract Registration
+
+A Nano Contract must be registered to have its transactions properly mapped. Transactions belonging to the wallet are already processed for balance purpose, however, they are not assigned to a specific nano contract by default, which may difficult query. By registering a Nano Contract we can mount a map of its transactions for easy access and also to process its balances.
```
Only the ones belonging to the user's wallet.

alexruzenhack replied:
Now we will show all Nano Contract's trasactions.

alexruzenhack commented on projects/hathor-wallet-mobile/0002-nano-contract-integration.md:
```diff
@@ -0,0 +1,591 @@
+
+A Nano Contract must be registered to have its transactions properly mapped. Transactions belonging to the wallet are already processed for balance purpose, however, they are not assigned to a specific nano contract by default, which may difficult query. By registering a Nano Contract we can mount a map of its transactions for easy access and also to process its balances.
+
+Users arrive at this screen from the *Main* screen to register a new Nano Contract. The screen features a form with an input field for *Nano Contract ID* and a *Register* button.
```
But why? Why we can't just identify all the addresses that have engaged with the Nano Contract being registered? This is possible, right? As I saw, all the user's transactions, including Nano Contract ones, are being loaded in the wallet.

alexruzenhack replied:
I understood.

alexruzenhack commented on projects/hathor-wallet-mobile/0002-nano-contract-integration.md:
```diff
@@ -0,0 +1,591 @@
+- On redux state `nanoContractsHistory`:
+	- It must be a map of `<TxHash, NcTxHistory>`.
+	- The `NcTxHistory` must contain a set of balance per token.
+	- The `NcTxHistory` must contain relevante information such as:
```
Added.

alexruzenhack commented on projects/hathor-wallet-mobile/0002-nano-contract-integration.md:
```diff
@@ -0,0 +1,568 @@
+
+This screen is dedicated to displaying a list of transactions associated with the selected *Nano Contract*. After a user selects a *Nano Contract* from the *[Nano Contract List](#component-nano-contract-list)* component, it is directed to this screen. From this screen a user can also inspect Nano Contract's **state** and **unregister** it.
+
+The header section of this screen provides basic information about the selected *Nano Contract* such as *Nano Contract ID*, *Blueprint Name* and *Registered Address*. In addition, the header also shows an action to **unregister** the Nano Contract and an action to see Nano Contract' **state** on explorer website.
```
I don't see any problem to have two tabs here. However, I believe the state screen deservers a careful thought. I mean, maybe we need a little design just for it, to consider its flexibility first. For now, can we stick with the explorer?

Regarding the address edition, I saw it on the design. I will increment the design and the implementation to consider it.

pedroferreira1 replied:
For me it's ok to discuss it separately, I agree it's the most complex screen. We just can't forget about it and Ana's Mockup must have it.

alexruzenhack commented on projects/hathor-wallet-mobile/0002-nano-contract-integration.md:
```diff
@@ -0,0 +1,568 @@
+- `network`: `mainnet`, `testnet`, `privatenet`
+- `stage`: `mainnet`, `testnet`
+
+Second, we should enable user's device in the `wallet-connect-mobile.rollout` feature flag on Unleash.
```
Do you think this part should be optional on nano contract activation?

Ok, it impacts the call to the nano contract RPC. If nano contract is inactive, it should not triggers the request modal.

pedroferreira1 replied:
Yes, that's the only UI/UX change that must be foreseen.

Nano contracts enabled and wallet connect disabled -> user can see data, register nano, read state, see history but can't interact with any dApp
Nano contracts disabled and wallet connect enabled -> user can't see any nano data, but can interact with dApps, except with nano contract calls.

pedroferreira1 commented on projects/hathor-wallet-mobile/0002-nano-contract-integration.md:
```diff
@@ -0,0 +1,568 @@
+
+Users arrive at this screen from the *Main* screen to register a new Nano Contract associated with a wallet's address. The screen features a form with an input field for *Nano Contract ID* and another input field for *Wallet's Address* and a *Register* button.
+
+The first wallet's address fills the *Wallet's Address* field by default. A user can change the address by tapping over *Wallet's Address* field that opens [*Wallet's Address List*](#modal-wallets-address-list) modal, containing a list of wallet's address with index and hash information, sorted by index in ascending order.
```
We are penalizing probably less than 0.1% of the users and preventing many users from making mistakes and requesting support from us.

That's the main reason I think we should allow users to change the registration address in the state screen. (Ana's support already reproduces this suggestion)

alexruzenhack replied:
> We are penalizing probably less than 0.1% of the users and preventing many users from making mistakes and requesting support from us.

Understood. I will need to change the implementation of history then, the way I store it using a compose key with address+nano_contract_id.
